`TemporalPropertyValueConverter` is called multiple times for Date `@Field` in Repository queries

Hello,

I believe I have come across a regression between `4.2.0` and `4.3.0` of the `spring-data-elasticsearch` library.

Please find attached a zip file containing two sample projects: one in 4.2.0 and the other in 4.3.0. To run it, first launch `docker-compose up` (from either one of the folders) and then try to execute the single test. 
[demo-temporalpropertyvalueconverter.zip](https://github.com/spring-projects/spring-data-elasticsearch/files/7650518/demo-temporalpropertyvalueconverter.zip)

The `4.2.0` test runs successfully, however the `4.3.0` runs into the following runtime exception:
```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.time.temporal.TemporalAccessor (java.lang.String and java.time.temporal.TemporalAccessor are in module java.base of loader 'bootstrap')
```
Looking into the stack trace and adding breakpoints, it looks like the `TemporalPropertyValueConverter#write()` method is called more than it should. It is first called to convert the `OffsetDateTime` field to the specified `@Field` format, but then it is called again to try to convert the already-converted string (e.g `2021-11-30T16:21:04.927+01:00`), causing a ClassCastException because a String is not a TemporalAccessor.

Moving up the debugger stack trace, it looks like that the `MappingElasticsearchConverter#updateQuery` method being called multiple times when executing Repository queries is what may be causing this issue?

I have not included the following test in the zip but Criteria queries (and Native ones) work fine:
```java
	@Test
	void criteria() {
		testDocumentRepository.save(new TestDocument(UUID.randomUUID(), OffsetDateTime.now()));
		Query query = new CriteriaQuery(new Criteria("startDate").between(OffsetDateTime.now().minusDays(3), OffsetDateTime.now().plusDays(1)));
		SearchHits<TestDocument> hits = elasticsearchOperations.search(query, TestDocument.class);
		Assertions.assertThat(hits.getTotalHits()).isNotZero();
	}
```

Thanks

--- a/src/main/java/org/springframework/data/elasticsearch/core/convert/AbstractRangePropertyValueConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/convert/AbstractRangePropertyValueConverter.java
@@ -76,7 +76,10 @@ public abstract class AbstractRangePropertyValueConverter<T> extends AbstractPro
 	public Object write(Object value) {
 
 		Assert.notNull(value, "value must not be null.");
-		Assert.isInstanceOf(Range.class, value, "value must be instance of Range.");
+
+		if (!Range.class.isAssignableFrom(value.getClass())) {
+			return value.toString();
+		}
 
 		try {
 			Range<T> range = (Range<T>) value;

--- a/src/main/java/org/springframework/data/elasticsearch/core/convert/DatePropertyValueConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/convert/DatePropertyValueConverter.java
@@ -60,6 +60,10 @@ public class DatePropertyValueConverter extends AbstractPropertyValueConverter {
 	@Override
 	public Object write(Object value) {
 
+		if (!Date.class.isAssignableFrom(value.getClass())) {
+			return value.toString();
+		}
+
 		try {
 			return dateConverters.get(0).format((Date) value);
 		} catch (Exception e) {

--- a/src/main/java/org/springframework/data/elasticsearch/core/convert/TemporalPropertyValueConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/convert/TemporalPropertyValueConverter.java
@@ -50,9 +50,9 @@ public class TemporalPropertyValueConverter extends AbstractPropertyValueConvert
 			try {
 				return dateConverter.parse(s, (Class<? extends TemporalAccessor>) actualType);
 			} catch (Exception e) {
+
 				if (LOGGER.isTraceEnabled()) {
 					LOGGER.trace(e.getMessage(), e);
-
 				}
 			}
 		}
@@ -63,6 +63,10 @@ public class TemporalPropertyValueConverter extends AbstractPropertyValueConvert
 
 	@Override
 	public Object write(Object value) {
+
+		if (!TemporalAccessor.class.isAssignableFrom(value.getClass())) {
+			return value.toString();
+		}
 
 		try {
 			return dateConverters.get(0).format((TemporalAccessor) value);

--- a/src/main/java/org/springframework/data/elasticsearch/core/mapping/PropertyValueConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/mapping/PropertyValueConverter.java
@@ -24,7 +24,8 @@ package org.springframework.data.elasticsearch.core.mapping;
 public interface PropertyValueConverter {
 
 	/**
-	 * Converts a property value to an elasticsearch value.
+	 * Converts a property value to an elasticsearch value. If the converter cannot convert the value, it must return a
+	 * String representation.
 	 *
 	 * @param value the value to convert, must not be {@literal null}
 	 * @return The elasticsearch property value, must not be {@literal null}

--- a/src/test/java/org/springframework/data/elasticsearch/core/convert/PropertyValueConvertersUnitTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/core/convert/PropertyValueConvertersUnitTests.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core.convert;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.params.provider.Arguments.*;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Named;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.springframework.data.elasticsearch.annotations.DateFormat;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.data.elasticsearch.core.mapping.PropertyValueConverter;
+import org.springframework.data.elasticsearch.core.mapping.SimpleElasticsearchMappingContext;
+import org.springframework.data.elasticsearch.core.mapping.SimpleElasticsearchPersistentEntity;
+import org.springframework.lang.Nullable;
+
+/**
+ * @author Peter-Josef Meisch
+ */
+public class PropertyValueConvertersUnitTests {
+
+	@ParameterizedTest(name = "{0}") // #2018
+	@MethodSource("propertyValueConverters")
+	@DisplayName("should return original object on write if it cannot be converted")
+	void shouldReturnOriginalObjectOnWriteIfItCannotBeConverted(PropertyValueConverter converter) {
+
+		NoConverterForThisClass value = new NoConverterForThisClass();
+
+		Object written = converter.write(value);
+
+		assertThat(written).isEqualTo(value.toString());
+	}
+
+	static Stream<Arguments> propertyValueConverters() {
+
+		SimpleElasticsearchMappingContext context = new SimpleElasticsearchMappingContext();
+		SimpleElasticsearchPersistentEntity<?> persistentEntity = context
+				.getRequiredPersistentEntity(NoConverterForThisClass.class);
+		ElasticsearchPersistentProperty persistentProperty = persistentEntity.getRequiredPersistentProperty("property");
+
+		List<PropertyValueConverter> converters = new ArrayList<>();
+
+		converters.add(new DatePropertyValueConverter(persistentProperty,
+				Collections.singletonList(ElasticsearchDateConverter.of(DateFormat.basic_date))));
+		converters.add(new DateRangePropertyValueConverter(persistentProperty,
+				Collections.singletonList(ElasticsearchDateConverter.of(DateFormat.basic_date))));
+		converters.add(new NumberRangePropertyValueConverter(persistentProperty));
+		converters.add(new TemporalPropertyValueConverter(persistentProperty,
+				Collections.singletonList(ElasticsearchDateConverter.of(DateFormat.basic_date))));
+		converters.add(new TemporalRangePropertyValueConverter(persistentProperty,
+				Collections.singletonList(ElasticsearchDateConverter.of(DateFormat.basic_date))));
+
+		return converters.stream().map(propertyValueConverter -> arguments(
+				Named.of(propertyValueConverter.getClass().getSimpleName(), propertyValueConverter)));
+	}
+
+	static class NoConverterForThisClass {
+		@SuppressWarnings("unused")
+		@Nullable Long property;
+	}
+}

--- a/src/test/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentPropertyUnitTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentPropertyUnitTests.java
@@ -265,23 +265,30 @@ public class SimpleElasticsearchPersistentPropertyUnitTests {
 
 	// region entities
 	static class FieldNameProperty {
-		@Nullable @Field(name = "by-name") String fieldProperty;
+		@Nullable
+		@Field(name = "by-name") String fieldProperty;
 	}
 
 	static class FieldValueProperty {
-		@Nullable @Field(value = "by-value") String fieldProperty;
+		@Nullable
+		@Field(value = "by-value") String fieldProperty;
 	}
 
 	static class MultiFieldProperty {
-		@Nullable @MultiField(mainField = @Field("mainfield"),
+		@Nullable
+		@MultiField(mainField = @Field("mainfield"),
 				otherFields = { @InnerField(suffix = "suff", type = FieldType.Keyword) }) String mainfieldProperty;
 	}
 
 	static class DatesProperty {
-		@Nullable @Field(type = FieldType.Date, format = {}, pattern = "dd.MM.uuuu") LocalDate localDate;
-		@Nullable @Field(type = FieldType.Date, format = DateFormat.basic_date_time) LocalDateTime localDateTime;
-		@Nullable @Field(type = FieldType.Date, format = DateFormat.basic_date_time) Date legacyDate;
-		@Nullable @Field(type = FieldType.Date, format = {}, pattern = "dd.MM.uuuu") List<LocalDate> localDateList;
+		@Nullable
+		@Field(type = FieldType.Date, format = {}, pattern = "dd.MM.uuuu") LocalDate localDate;
+		@Nullable
+		@Field(type = FieldType.Date, format = DateFormat.basic_date_time) LocalDateTime localDateTime;
+		@Nullable
+		@Field(type = FieldType.Date, format = DateFormat.basic_date_time) Date legacyDate;
+		@Nullable
+		@Field(type = FieldType.Date, format = {}, pattern = "dd.MM.uuuu") List<LocalDate> localDateList;
 	}
 
 	static class SeqNoPrimaryTermProperty {
@@ -344,8 +351,10 @@ public class SimpleElasticsearchPersistentPropertyUnitTests {
 
 	private static class EntityWithCustomValueConverters {
 		@Id private String id;
-		@Nullable @ValueConverter(ClassBasedValueConverter.class) private String fieldWithClassBasedConverter;
-		@Nullable @ValueConverter(EnumBasedValueConverter.class) private String fieldWithEnumBasedConverter;
+		@Nullable
+		@ValueConverter(ClassBasedValueConverter.class) private String fieldWithClassBasedConverter;
+		@Nullable
+		@ValueConverter(EnumBasedValueConverter.class) private String fieldWithEnumBasedConverter;
 	}
 
 	private static class ClassBasedValueConverter implements PropertyValueConverter {

